### ISSUE TITLE
Readd The Download Time Bar

### ISSUE THREAD

**Issue by Cernelius** [triplea-game/triplea#1807]
Now that I'm downloading Iron War with 1.9.0.0.4520 I see that there is not anymore the bar showing you the download progress (the percentage of completed download, giving you a hint of how much you have to wait).
I don't know why it has been removed, but I liked that addition, especially for heavy (over 100 MB) maps; so I suggest to have it back, if no major reasons against it (like if it just made the download a bit slower?).
Feel free to just close this issue and redirect where the matter has been already discussed / decided.

**Comment by Cernelius** [comment]
I see the 100% bar once the download is finished, but I don't see it while in progress anymore. Downloading Iron War took more than 30 minutes for me (I see it is 340MB), so it would have been surely nice to have an idea. I'm thinking mostly in term of people in lobby that are just downloading a map to play with someone else (maybe after crashing because they don't have it), thus would be nice for them (and their mates!) having a clue of how much they are going to make the other people wait for them, while they are downloading the map.

**Comment by ssoloff** [comment]
@Cernelius So you don't see the following indeterminate progress bar at all?  The entire lower pane is empty until the download is complete?

![iron-war-download](https://cloud.githubusercontent.com/assets/4826349/26764199/91a07cb2-4930-11e7-978b-7add4607f157.png)

I can possibly explain why you don't see any progress until 100%.  Occasionally (about 5-10% of the time in my experience), GitHub does not tell us the length of a download.  If we don't have the length, we can't display the progress of the download as a percentage.  The addition of the indeterminate state to the progress bar was an enhancement to capture this case.  Otherwise, it would just sit at 0% and then suddenly jump to 100% when the download was complete.  The indeterminate animation was intended to give the user a "feeling" that something was happening even if we can't report otherwise.

However, I am concerned if you don't see the progress bar at all until the download is complete.  Does this happen consistently for you, or only when you attempted to recently download Iron War?  Can you please try a few other (smaller) downloads to confirm?

**Comment by ssoloff** [comment]
One more question...  Were you **installing** Iron War or **updating** it?

**Comment by Cernelius** [comment]
I had never downloaded Iron War before.
I was seeing that indeterminate progress bar, but not seeing the percentage progress bar.
Does GitHub not tell the length for some reasons, like if a map is bigger than some size or it has an item in it that it is bigger than a max limit? **Is there a way for a mapmaker to know what it takes to have the bar for its map?**
Otherwise, I guess this issue can be closed.

I'm currently hosting some bots; so I'm not going to redownload stuff. I don't remember to have ever seen a not percentage bar since 1.9 was released, so I assumed you removed this feature at some point.

**Comment by ssoloff** [comment]
The percentage progress bar still exists; it has not been removed.  The only feature that is new is the addition of the indeterminate progress bar.

Ideally, you should see the following sequence:

* Indeterminate progress bar displayed for a second or two (until we receive the download length)
* Percentage progress bar displayed at 0%
* Percentage progress bar gradually increases as the download proceeds
* Percentage progress bar displayed at 100%

However, as I said, if we do not receive the download length, you will instead see the following sequence:

* Indeterminate progress bar displayed until the download is complete
* Percentage progress bar displayed at 100%

When GitHub decides to not send us the content length, it is because it is in the process of dynamically generating the map zipball we requested and does not yet know the final length.  (Map makers have no control over this.)  So instead of waiting until the zipball is fully created, it starts to stream the "front" of the file to us while it continues to add data to the "back."  They call this a "chunked" transfer because the file is sent in chunks, and we don't know the length of each chunk until we completely receive it.

The map zipballs we download from GitHub are theoretically dynamically generated every time someone requests one.  However, in reality, once a zipball is created, GitHub probably caches it until it detects a new commit in the repo (e.g. Frostion updating something in the Iron War map) or until a set period of time has passed (e.g. no one has requested to download the zipball in 12 hours).

You were unfortunately in the window where GitHub had to regenerate the zipball.  If someone else tried to download the same map a minute or two later (presumably the zipball can be created on the order of seconds), they would have received the download length and seen the progress bar normally (as I did when I generated the screenshot above).

**Comment by Cernelius** [comment]
Thanks for the detailed explanation. Duno where this might be copy-pasted, for reference, in case anyone might wonder in the future, about this stuff.

**Comment by ssoloff** [comment]
Yeah, I'm not sure either where the best place to document it is.

Note that there has been a discussion recently of changing where the maps are hosted (#1772), so this problem may go away if we use a more traditional method.
